samples/triangle/main.c

```diff
@@ -0,0 +1,281 @@
+    {{-1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}},
+    {{-1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{ 1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{-1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}},
+    {{ 1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{ 1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}},
+    {{-1.0, -1.0,  1.0}, { 1.0,  0.0,  0.0}},
+    {{ 0.0,  1.0,  1.0}, { 0.0,  1.0,  0.0}},
+    {{ 1.0, -1.0,  1.0}, { 0.0,  0.0,  1.0}},
+};
+
+#define MASK(mask, val) (((val) << (ffs(mask)-1)) & (mask))
+
+static void matrix_viewport(float out[4][4], float x, float y, float width, float height, float z_min, float z_max);
+static void init_shader(void);
+static void set_attrib_pointer(unsigned int index, unsigned int format, unsigned int size, unsigned int stride, const void* data);
+static void draw_arrays(unsigned int mode, int start, int count);
+
+/* Main program function */
+void main(void)
+{
+    uint32_t *p;
+    int       i, status;
+    int       width, height;
+    int       start, last, now;
+    int       fps, frames, frames_total;
+
+    if ((status = pb_init())) {
+        debugPrint("pb_init Error %d\n", status);
+        XSleep(2000);
+        XReboot();
+        return;
+    }
+
+    pb_show_front_screen();
+
+    /* Basic setup */
+    width = pb_back_buffer_width();
+    height = pb_back_buffer_height();
+
+    /* Load constant rendering things (shaders, geometry) */
+    init_shader();
+    alloc_vertices = MmAllocateContiguousMemoryEx(sizeof(verts), 0, 0x3ffb000, 0, 0x404);
+    memcpy(alloc_vertices, verts, sizeof(verts));
+    num_vertices = sizeof(verts)/sizeof(verts[0]);
+    matrix_viewport(m_viewport, 0, 0, width, height, 0, 65536.0f);
+
+    /* Setup to determine frames rendered every second */
+    start = now = last = XGetTickCount();
+    frames_total = frames = fps = 0;
+
+    while(1) {
+        pb_wait_for_vbl();
+        pb_reset();
+        pb_target_back_buffer();
+
+        /* Clear depth & stencil buffers */
+        pb_erase_depth_stencil_buffer(0, 0, width, height);
+        pb_fill(0, 0, width, height, 0x00000000);
+        pb_erase_text_screen();
+
+        while(pb_busy()) {
+            /* Wait for completion... */
+        }
+
+        /* Send shader constants
+         *
+         * WARNING: Changing shader source code may impact constant locations!
+         * Check the intermediate file (*.inl) for the expected locations after
+         * changing the code.
+         */
+        p = pb_begin();
+
+        /* Set shader constants cursor at C0 */
+        pb_push1(p, NV20_TCL_PRIMITIVE_3D_VP_UPLOAD_CONST_ID, 96); p+=2;
+
+        /* Send the transformation matrix */
+        pb_push(p++, NV20_TCL_PRIMITIVE_3D_VP_UPLOAD_CONST_X, 16);
+        memcpy(p, m_viewport, 16*4); p+=16;
+
+        pb_end(p);
+        p = pb_begin();
+
+        /* Clear all attributes */
+        pb_push(p++,NV097_SET_VERTEX_DATA_ARRAY_FORMAT,16);
+        for(i = 0; i < 16; i++) {
+            *(p++) = 2;
```

Comment:
Not quite sure. This comes from graphics.c@clear_attribs. NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_F maybe?